docs/proposals/20210331-cascadingdeletionprotection.md

```diff
@@ -0,0 +1,97 @@
+
+## Motivation
+
+Currently, there are so many risks can be caused by cascading deletion mechanism:
+
+1. delete a CRD mistakenly, all CR disappeared
+2. delete a Workload mistakenly, all Pods belongs to it deleted
+3. delete all Workloads mistakenly in batches, all applications in the cluster unavailable
+4. delete a namespace, all resources in it deleted
+
+Kruise should provide a safety policy which could help users protect Kubernetes resources and
+applications' availability from the cascading deletion mechanism.
+
+## Proposal
+
+API definition:
+
+- a new feature-gate named `ResourcesDeletionProection`
+- a label named `policy.kruise.io/delete-protection`, the values can be:
+  - `Always`: this object will always be forbidden to be deleted, unless the label is removed
+  - `Cascading`: this object will be forbidden to be deleted, if it has active resources owned
+
+If the feature-gate has enabled, the resources below with this label will be validated for deletion operation:
+
+| Kind                        | Group                  | Version            | Cascading judgement                                |
+| --------------------------- | ---------------------- | ------------------ | ----------------------------------------------------
+| `Namespace`                 | core                   | v1                 | whether there is active Pods in this namespace     |
+| `CustomResourceDefinition`  | apiextensions.k8s.io   | v1beta1, v1        | whether there is existing CRs of this CRD          |
+| `Deployment`                | apps                   | v1                 | whether the replicas is 0                          |
+| `StatefulSet`               | apps                   | v1                 | whether the replicas is 0                          |
+| `ReplicaSet`                | apps                   | v1                 | whether the replicas is 0                          |
+| `CloneSet`                  | apps.kruise.io         | v1alpha1           | whether the replicas is 0                          |
+| `StatefulSet`               | apps.kruise.io         | v1alpha1, v1beta1  | whether the replicas is 0                          |
+| `UnitedDeployment`          | apps.kruise.io         | v1alpha1           | whether the replicas is 0                          |
+
+### User Stories
+
+1. delete a CRD mistakenly, all CR disappeared
+2. delete a Workload mistakenly, all Pods belongs to it deleted
+3. delete all Workloads mistakenly in batches, all applications in the cluster unavailable
+4. delete a namespace, all resources in it deleted
+
+### Requirements
+
+If users enable `ResourcesDeletionProection` feature-gate when install or upgrade Kruise,
+Kruise will require more authorities:
+
+1. Webhook for deletion operation of namespace, crd, deployment, statefulset, replicaset and workloads in Kruise.
+2. ClusterRole for reading all resource types, because CRD validation needs to list the CRs of this CRD.
```

Comment:
Is it possible to add a feature gate for CustomResourceDefinition protection? privilege for all resource types are just so huge. we can protect the kruise crd only by default